Moved the logic to gate WPCOM blocks to a shared file

pablinos opened the pull request:
#### Changes proposed in this Pull Request:

The logic to gate the blocks to paid plans on WPCOM is repeated for the
OpenTable and Calendly blocks. It is likely to be needed in future too,
so this change moves the repeated to logic to the `Jetpack_Gutenberg`
library.

This may not be the best place for this code to live, so opinions are welcome!

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
This is a refactor, so changes an existing feature.

#### Testing instructions:
* Enable beta blocks on a test site
* Check that you can still add and use the OpenTable block

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
No entry needed

pablinos committed:
[not verified] [not verified] # This is a combination of 5 commits.

[not verified] Moved the logic to gate WPCOM blocks to a shared file

The logic to gate the blocks to paid plans on WPCOM is repeated for the
OpenTable and Calendly blocks. It is likely to be needed in future too,
so this change moves the repeated to logic to the `Jetpack_Gutenberg`
library.

Reworked incorporating ideas from #14430 and #14431

Changed availability logic so a block is always registered

Based on the suggestion by @simison, this changes the logic around block
availability, so that a block can always be registered, but its
availability can be set independently and used to prevent dynamic blocks
from rendering.

Adjusted the call to `remove_extension_prefix` because of load order

The compatibility function `wp_startswith` isn't loaded until the
`plugins_loaded` hook, but plugins sometimes call
`jetpack_register_block` before that. This caused an exception in the
`remove_extension_prefix` call.

This moves that call to the point just before the value is used,
ensuring that the function is available.

Refactored to work with the existing `Jetpack_Plan::supports` method

This is missing the WPCOM part at the moment.

scruffian commented on class.jetpack-gutenberg.php:
```diff
@@ -35,6 +35,19 @@ function jetpack_register_block( $slug, $args = array() ) {
 
+/**
+ * Registers a gutenberg block for WPCOM users with a
+ * pard pland and all Jetpack users
```
what's a `pard pland`? :)

jeherve commented on class.jetpack-gutenberg.php:
```diff
@@ -801,4 +814,49 @@ public static function get_extensions_preset_for_variation( $preset_extensions_m
+				$slug,
+				'missing_plan',
+				array(
+					'required_feature' => 'opentable',
```
This would probably need to be dynamic.

pablinos replied:
Oops! Yes, definitely, I'll fix that.

jeherve commented on class.jetpack-gutenberg.php:
```diff
@@ -801,4 +814,49 @@ public static function get_extensions_preset_for_variation( $preset_extensions_m
+	 *
+	 * @return bool
+	 */
+	public static function is_premium_wpcom_available() {
```
Maybe we should make this a bit more generic and have a general "paid" block check where we would pass the minimum plan needed for availability? This way we could use this in the future for other blocks that may require a different plan, without having to create a new function just for that.

pablinos replied:
Yes, I was wondering about how generic to make it and what to call it as well.

simison commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -12,52 +12,28 @@
-			)
-		);
-	}
+	jetpack_register_premium_wpcom_block(
```
Could the plan requirement information be passed to `jetpack_register_block` instead of a new separate function?

I don't have strong opionion on this but just seems easier to maintain one function.

simison replied:
It also feels little confusing to have "wpcom" in the registration name, isn't this block available for Jetpack sites?

simison commented:
One thing to note is how availability for recurring payments and simple payments blocks are registered differently depending on if we're on wpcom or jetpack, example:

https://github.com/Automattic/jetpack/blob/aa3ad69b50607bda893d7a2b22bf74ae3c81c932/modules/memberships/class-jetpack-memberships.php#L304-L317

jeherve commented:
I've been thinking about this some, and liked @simison's idea here:
https://github.com/Automattic/jetpack/pull/14426#discussion_r369485685

I thought I'd see how this could look like, so here is my take: #14431

scruffian commented:
I had some ideas here too: https://github.com/Automattic/jetpack/pull/14430

pablinos commented:
I've incorporated some ideas from both of your PRs (@scruffian and @jeherve ).

I'm not sure if things should live where they currently are. I noticed that the `Jetpack_Plan` class deals with all the WPCOM plan slugs too, so I decided to add our helper functions in there. It might be more sensible to create general `Plan_Helpers` lib and get `Jetpack_Plan` to use it.

I also decided to pass the `required_plans` in the args array for `jetpack_register_block`. The thinking was that it keeps the signature the same and we can add other arguments this way if we want to, in the future.

It also needs some testing on WPCOM, but it seems to be working in the manual tests I've done on my Jetpack site. I'll add some unit tests too.

scruffian commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -9,55 +9,34 @@
-}
+const FEATURE_NAME   = 'opentable';
+const BLOCK_NAME     = 'jetpack/' . FEATURE_NAME;
+const REQUIRED_PLANS = array( 'wpcom' => 'premium-plan' );
```
IMO it would be valuable to generalize this slightly more. We might want to hide blocks for all users on either Jetpack or WPCOM. If we made both keys (`jetpack` and `wpcom`) required, then we could add either `true` or `false`, to simply enable/disable the block in these environments.

pablinos replied:
Yeah, that's a good idea. We could leave it that you don't have to provide the key, but set it to `false` to turn them off because the blocks are available by default.

scruffian commented on class.jetpack-plan.php:
```diff
@@ -153,16 +298,9 @@ public static function get() {
 		}
 
 		// Define what paid modules are supported by professional plans.
```
Should this comment say "business"?

pablinos replied:
I'm not sure, but looking at the code that follows, I would say so, yes.

kraftbj replied:
Professional and Business are the same thing. On WP.com, it is called "Business", in Jetpack it is called "Professional".

scruffian commented on class.jetpack-plan.php:
```diff
@@ -23,6 +23,163 @@ class Jetpack_Plan {
+	 * @access public
+	 * @static
+	 *
+	 * @return bool True if the site is a WordPress.com Simple Site.
```
What about Atomic?

pablinos replied:
Is `IS_WPCOM` defined for Atomic? I'll have to check that.

jeherve replied:
It's not. `IS_WPCOM` is meant to check for simple sites. For all intents and purposes, Atomic sites are Jetpack sites.

If we do need to check for Atomic at any point, this can help:
https://github.com/Automattic/jetpack/blob/31efa189ad223c0eb7ad085ac0650a23facf9ef5/functions.global.php#L75

scruffian commented:
I like the concepts here but I'm not sure this approach makes sense for Jetpack - Jetpack already has a `supports` function and arrays which contain features. I would anticipate that for Jetpack rather than supply a slug we could just call the `supports` function like:

`Jetpack_Plan::supports( BLOCK_NAME )`

which might simplify things a bit.

simison commented:
> I also decided to pass the required_plans in the args array for jetpack_register_block. The thinking was that it keeps the signature the same and we can add other arguments this way if we want to, in the future.

Thinking this again, we've had two separate functions: one that registers the block and another that sets block available or unavailable (with information about the reason).

Plan restriction seems like a relevant bit for setting the availability, not for registering the block. I.e. the block should always be registered.

pablinos commented:
> Plan restriction seems like a relevant bit for setting the availability, not for registering the block. I.e. the block should always be registered.
> 
> What do you think?

Yes, I agree with that, but [the way the availability works](https://github.com/Automattic/jetpack/blob/c9e2901cd559ec75322abd21a50d6e00eddbee78/class.jetpack-gutenberg.php#L392), if a block is registered it can't be made unavailable, which is why the other blocks have done it this way I think. 

We could change that availability logic of course!

pablinos commented:
> I would anticipate that for Jetpack rather than supply a slug we could just call the `supports` function like:
> 
> `Jetpack_Plan::supports( BLOCK_NAME )`
> 
> which might simplify things a bit.

Maybe, yes, but that does complicate setting things up. You need to add the feature to the plans, and it's a different check as part of the availability logic. In practice, I see it that features will be available on a tier of plan and higher, as they are on WPCOM.

But as you say, there is already a system in place that's more granular, and this introduces another mechanism. Either way has pros and cons, so it probably needs some more opinions 😄

scruffian commented:
We also have a similar mechanism on WPCOM for feature availability checks in the `Store_Product_List` class. We _could_ move this file into Jetpack so its accessible in both places, and then use it to check block availability. 

This approach makes the most sense to me, but it's more political.

pablinos commented:
> Plan restriction seems like a relevant bit for setting the availability, not for registering the block

Actually there's another part to this. Availability of the block only gets checked in the editor itself, and for server side rendered blocks we need the block to only be registered, when it's available.

We could check in the render callback function if the block is available? We could even wrap that function as the block is registered and return if it's unavailable. So many options!

pablinos commented:
Ok, hold that thought, I don't think this approach is going to work. I'm trying something else...

pablinos committed:
[not verified] Added the first version of the plan check for WPCOM

pablinos committed:
[not verified] Stop the availability check for non-jetpcak blocks

A core test was failing for core/video because Videopress registers a
server render_callback. This is the exception and we don't want to do
avaavailability checks on non-jetpack blocks, so this change adds a
check on the slug as the block is registered.

pablinos committed:
[not verified] Corrected the comments on the availability methods

pablinos committed:
[not verified] Updated logic to supply minimum plan slug for the WPCOM nudge

pablinos commented:
I think this is probably ready for another look. I'm taking the approach of checking if a plan has support for a particular block. It now also incorporates @simison's suggestion of blocks always being registered, but their availability being set as required.

I'll have a look at simplifying the code in places as CodeClimate is rightly complaining, but the logic seems to work so far.

I've been manually updating D38154-code, and I've added some additional changes there to include the `opentable` feature on Premium and higher plans.

scruffian commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -55,18 +32,7 @@ function register_block() {
-			)
-		);
-	}
+	\Jetpack_Gutenberg::set_availability_for_plan( FEATURE_NAME );
```
I wonder if we should use the block name, rather than the feature name, so that we can support premium blocks from other places in WPCOM?

pablinos replied:
We could but the availability array that's created uses the module/feature name as the key, so it's following that pattern for now. We could make some further changes to the availibility code, but that's probably for another PR.

scruffian commented on class.jetpack-gutenberg.php:
```diff
@@ -801,4 +815,46 @@ public static function get_extensions_preset_for_variation( $preset_extensions_m
+			if ( ! class_exists( 'Store_Product_List' ) ) {
+				require WP_CONTENT_DIR . '/admin-plugins/wpcom-billing/store-product-list.php';
+			}
+			$features_data = Store_Product_List::get_site_specific_features_data();
```
Should we pass in the site_id, or if this isn't site specific then can we just use `get_feature_list`

pablinos replied:
We can, but the default is to call `get_current_blog_id()` if the ID isn't passed in. It is the site-specific data we need as it gives a collection of active features and available ones, including the plans that the site would need to upgrade to.

scruffian commented on class.jetpack-gutenberg.php:
```diff
@@ -801,4 +815,46 @@ public static function get_extensions_preset_for_variation( $preset_extensions_m
+				$slug,
+				'missing_plan',
+				array(
+					'required_feature' => self::remove_extension_prefix( $slug ),
```
Since we're passing in the block name without the `jetpack/` do we need this?

pablinos replied:
Yes, I think I was doing it more as a safeguard. You're right, it's not needed!

scruffian commented on class.jetpack-gutenberg.php:
```diff
@@ -801,4 +815,46 @@ public static function get_extensions_preset_for_variation( $preset_extensions_m
+		} elseif ( ! jetpack_is_atomic_site() ) {
+			/*
+			 * If it's Atomic then assume all features are available
+			 * otherwise check against the Jetpack plan.
```
Should we add this logic in now?

pablinos replied:
As discussed, for now, we'll assume that features are available on Atomic sites.

scruffian commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -55,18 +32,7 @@ function register_block() {
-	}
+	\Jetpack_Gutenberg::set_availability_for_plan( FEATURE_NAME );
 }
 add_action( 'jetpack_register_gutenberg_extensions', 'Jetpack\OpenTable_Block\set_availability' );
```
Should we make this part of every block, rather than having to call it separately?

pablinos replied:
Yes, I was thinking about that but as this is primarily regarding gating features to plans I thought that it should be left as a check that can be used for those blocks. There's some other availability logic in some of the blocks, so this could be used in conjunction with that.

On the other hand, if the majority of blocks will need a plan check, then we could register this hook as the blocks are registered, and then remove it on the others. Another option might be to pass another argument to `jetpack_register_block` that could indicate the check is needed.

That last option might be the cleanest. What do you think about that?

scruffian replied:
I think lets keep it explicit right now and revisit if we find we're copy/pasting this code a lot.

pablinos replied:
Sounds good!

scruffian commented on class.jetpack-gutenberg.php:
```diff
@@ -32,6 +32,19 @@ function jetpack_register_block( $slug, $args = array() ) {
 	}
 
+	// If the block is dynamic, and a Jetpack block, wrap the render_callback to check availability.
+	if ( 0 === strpos( $slug, 'jetpack/' ) && isset( $args['render_callback'] ) ) {
```
This is a bit hard to grok, maybe it should be in a different function?

pablinos replied:
Definitely, I'll extract it and try to make it a bit more readable.

scruffian replied:
Looks way better now :P)

scruffian commented:
This is looking really good. I just had a few minor questions. 👏

pablinos committed:
[not verified] Simplified the code to wrap the `render_callback`

jeherve commented:
This is a tough one. I wonder if this is going to be easy to understand for any new contributors. We go from having to check one place (the file where the block is registered) to find out its availability to having to check 3 (the file where the block is registered to see if `set_availability_for_plan` is used, `class.jetpack-plan.php` for Jetpack availability, and wpcom for wpcom availability). 
It seems like a lot, and may be too complex overall?

I think it may also be confusing for some that some free Jetpack blocks will end up being listed as free blocks in `class.jetpack-plan.php` (because their wpcom equivalent is not free) while other free Jetpack blocks won't be listed there (because they're free everywhere).

> blocks always being registered, but their availability being set as required.

I like this approach. 

Ideally, I think I'd personally like to see block registration handled from one function where I would provide all the details needed to render a registered block available or not, without having to rely on other files to define availability. I feel that keeping everything in one place may be easier to maintain on the long term, instead of having to maintain 2 lists of feature supports separated from the actual features. 
I think that's something you touched on here as well:
https://github.com/Automattic/jetpack/pull/14426#discussion_r372651900

I'd love to have feedback from different folks who developed the first availability features in Jetpack. @Automattic/luna Do you think you could take a look and let us know what you think?

pablinos commented:
Thanks @jeherve. I agree this is tricky. A lot of the problem comes down to there not being a single source of information for plan features, and a single mechanism for checking their availability across all environments.

> the file where the block is registered to see if set_availability_for_plan is used

Do you think it would help to always use `set_availability_for_plan` and explicitly state that a feature is available on the free plans?

jeherve commented:
> Do you think it would help to always use set_availability_for_plan

I think so, yeah

> and explicitly state that a feature is available on the free plans?

Maybe the method should only be used internally, when one specifies a specific parameter in `jetpack_register_block`? So when working on a new block you wouldn't have to worry about availability, you would only worry about specifying plan restrictions when registering the block, and behind the scenes Jetpack would use `set_availability_for_plan` to handle availability of the registered block?

sirreal commented:
> I'd love to have feedback from different folks who developed the first availability features in Jetpack. @Automattic/luna Do you think you could take a look and let us know what you think?

My head is in a very different space right now and I haven't been involved much in this project recently, so most of the details are not very fresh for me. In theory this seems like a reasonable change to make and no reasons to avoid this type of change jump out at me.

kraftbj commented:
I like this overall and it seems to work well. It'll need a rebase.

One thing I'd suggest is documenting the heck out of this, in particular -- if I produce a new block that is for a plan, how do I do it?

Perhaps add details at https://github.com/Automattic/jetpack/blob/master/extensions/README.md ?

apeatling commented:
Will rebase and retest this one.

apeatling commented:
✅Rebased

apeatling commented:
Tested and confirmed this is still working locally. I see beta blocks and can insert them on a local install with beta blocks setting enabled. If I sync to WordPress.com where the setting is not enabled, I do not see the blocks.

lancewillett commented:
What's the next step to unblock this one? More testing, or code feedback?

scruffian commented:
I think the main thing is some more documentation. I think it's on @pablinos's list but he's super snowed under at the moment.

pablinos commented:
I need to investigate @jeherve's [suggestion above](https://github.com/Automattic/jetpack/pull/14426#issuecomment-586305579) too. It makes sense to only specify plans when they're required.

creativecoder commented:
@pablinos will leave a checklist here of next steps required to move this forward. We need to pause this one for now, so we can re-start it when we're able to come back to it in a couple cycles.

pablinos committed:
Rebased to include the frontend upgrade nudge

This also updates the logic so the availability check is done in
`jetpack_register_block`

pablinos committed:
Change call to wp_startswith to allow the function to be called in init

The wp_startswith function isn't available on wpcom until
plugins_loaded. This changes the calls to it, to the strpos equivalent.

pablinos committed:
Move the availability check so it doesn't happen in the init hook

On wpcom, the availability check was getting the plans form the databse
and that was causing havoc with the isolated tests. I think this should
fix it.

pablinos commented:
I've rebased this and incorporated the changes from #14953. I had some significant trouble with testing on wpcom, but in the end, I think it was mainly a case of my sandbox not being up to date.

I've also made an attempt at @jeherve's suggestion (or at least my interpretation of it!). The availability checks are only performed if you pass `'plan_check'=>true` as an argument to `jetpack_register_block`.

In terms of what's left to do. If, as it stands, this change is ok, I'll need to:

- Add some documentation. In order to add a new paid block, it needs to be registered with the `plan_check` parameter and added to `Jetpack_Plan` and the wpcom products. I've updated D38154-code with those changes for the OpenTable and Calendly blocks.
- Refactor the other paid blocks to use this approach. I'm not sure how many there are, but potentially Simple Payments, Videopress, Recurring Payments.
- Work out what to do with Videopress. This is all based on blocks being named `jetpack/`, but Videopress overrides the core video block. We may need to tweak the approach there.

pablinos commented:
> Yes, I agree with that, but the way the availability works, if a block is registered it can't be made unavailable, which is why the other blocks have done it this way I think.
>
> We could change that availability logic of course!

It seems to be working the way it is at the moment - potentially because we set the avaliability before registering the block - but it would be good to check if we need to [change this logic too](https://github.com/Automattic/jetpack/blob/c9e2901cd559ec75322abd21a50d6e00eddbee78/class.jetpack-gutenberg.php#L392)

scruffian commented on class.jetpack-gutenberg.php:
```diff
@@ -165,8 +181,8 @@ private static function prepend_block_prefix( $block_name ) {
-	private static function remove_extension_prefix( $extension_name ) {
-		if ( wp_startswith( $extension_name, 'jetpack/' ) || wp_startswith( $extension_name, 'jetpack-' ) ) {
+	public static function remove_extension_prefix( $extension_name ) {
+		if ( 0 === strpos( $extension_name, 'jetpack/' ) || 0 === strpos( $extension_name, 'jetpack-' ) ) {
```
I'm not sure why you made this change.

pablinos replied:
Neither am I now, but it was something to do with `wp_startswith` not being available in some context. I'll change it back and see if I can reproduce the error

scruffian commented on class.jetpack-gutenberg.php:
```diff
@@ -458,8 +474,9 @@ public static function get_availability() {
 		foreach ( self::$extensions as $extension ) {
-			$is_available = self::is_registered( 'jetpack/' . $extension ) ||
-			( isset( self::$availability[ $extension ] ) && true === self::$availability[ $extension ] );
+			$is_set       = isset( self::$availability[ $extension ] );
```
I did a similar thing in https://github.com/Automattic/jetpack/pull/15665

scruffian reviewed (dismissed):
This looks like a great step in the right direction. I think we should merge it.

jeherve reviewed (changes requested):
This seems to work well in my tests on a Jetpack site. I have not tested on WordPress.com yet.

As I mentioned in #15665, I think we'll need to document all this so it can be easily understood by new contributors, or even by us a year from now :)

Ideally we'd have a section in `extensions/README.md` where a contributor could go to find out where in Jetpack and on WordPress.com you should add plan requirements whenever you create a new block.

pablinos commented on class.jetpack-gutenberg.php:
```diff
@@ -37,6 +37,22 @@ function jetpack_register_block( $slug, $args = array() ) {
 
+	$feature_name = Jetpack_Gutenberg::remove_extension_prefix( $slug );
+	// If the block is dynamic, and a Jetpack block, wrap the render_callback to check availability.
+	if ( isset( $args['plan_check'] ) && true === $args['plan_check'] && 0 === strpos( $slug, 'jetpack/' ) && isset( $args['render_callback'] ) ) {
```
That seems better to me! I suppose any block registered using this function should be a Jetpack block. I thing the `jetpack/` check is a hangover from a previous iteration.

pablinos committed:
Simplify the plan_check logic

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

pablinos commented on class.jetpack-gutenberg.php:
```diff
@@ -971,4 +988,66 @@ public static function upgrade_nudge( $plan ) {
+			}
+			$features_data = Store_Product_List::get_site_specific_features_data();
+			$is_available  = in_array( $slug, $features_data['active'], true );
+			if ( isset( $features_data['available'][ $slug ] ) && ! empty( $features_data['available'][ $slug ] ) ) {
```
I often make that mistake. Thanks for catching it!

pablinos committed:
Simplify the 'if' statement

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

pablinos commented:
> Ideally we'd have a section in extensions/README.md where a contributor could go to find out where in Jetpack and on WordPress.com you should add plan requirements whenever you create a new block.

Yes, that makes sense. I'll draft something and add it to the PR.

scruffian committed:
Add some documentation

scruffian commented:
@jeherve @pablinos I had a go at adding some documentation. Let me know what you think!

pablinos committed:
Added an extra line to the docs about where to find plan data.

pablinos commented:
Thanks @scruffian, it looks good to me. I've added a line to say where to make the changes to the plan data, referencing D43206-code, which should probably be merged before the WPCOM version of this.

jeherve commented on extensions/README.md:
```diff
@@ -154,6 +154,47 @@ rsync -az --delete _inc/blocks/ \
+
+The plan data is found in `class.jetpack-plan.php` for Jetpack and an example of adding the features to WordPress.com plans is in D43206-code.
+
+### Upgrades
+Paid blocks that aren't supported by a user's plan will still be registered for use in the block editor, but won't render on the frontend of the site. An `UpgradeNudge` component will display above the block in the editor and one the front end of the site to inform users that this is a paid block.
```
```suggestion
### Upgrades
Paid blocks that aren't supported by a user's plan will still be registered for use in the block editor, but will not be displayed by default.

You can, however, use the following filter:
```php
add_filter( 'jetpack_block_editor_enable_upgrade_nudge', '__return_true' );
```

This will allow you to take advantage of those registered blocks. They will not be rendered to logged out visitors on the frontend of the site, but the block will be available in the block picker in the editor. When you add a paid block to a post, an `UpgradeNudge` component will display above the block in the editor and on the front end of the site to inform users that this is a paid block.
```

jeherve reviewed (changes requested):
This tests well for me in Jetpack, but I can't see the paid blocks on WordPress.com simple sites with a level above the required one. 
See p1589467803004700-slack-explorers

I have suggested small documentation changes below, I'll apply them now. I'll also rebase all this while I am at it.

jeherve committed:
Documentation updates

jeherve reviewed (approved):
My bad, I was missing a patch. It works now. This should be good to merge!

scruffian commented:
r207476-wpcom
